### ISSUE TITLE
Throwing exception on simple use case

### ISSUE THREAD

**Issue by westhomas** [node-inspector/node-inspector#905]
First time using node-inspector. Here's my setup:

```
$ nvm --version
0.31.4
$ nvm current
v6.4.0
$ node --version
v6.4.0
$ npm install -g node-inspector
```

My simple test file:

```
//test.js
var d = {asdf: 123}
debugger;
console.log(d)
```

Firing up the debugger throws exception:
```
$ node-debug test.js
Node Inspector v0.12.8
Visit http://127.0.0.1:8080/?port=5858 to start debugging.
Debugging `test.js`

Debugger listening on [::]:5858
/Users/wes/.nvm/versions/node/v6.4.0/lib/node_modules/node-inspector/lib/InjectorClient.js:111
      cb(error, NM[0].ref);
                     ^

TypeError: Cannot read property 'ref' of undefined
    at InjectorClient.<anonymous> (/Users/wes/.nvm/versions/node/v6.4.0/lib/node_modules/node-inspector/lib/InjectorClient.js:111:22)
    at /Users/wes/.nvm/versions/node/v6.4.0/lib/node_modules/node-inspector/lib/DebuggerClient.js:121:7
    at Object.create.processResponse.value (/Users/wes/.nvm/versions/node/v6.4.0/lib/node_modules/node-inspector/lib/callback.js:23:20)
    at Debugger._processResponse (/Users/wes/.nvm/versions/node/v6.4.0/lib/node_modules/node-inspector/lib/debugger.js:95:21)
    at Protocol.execute (_debugger.js:121:14)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:177:18)
    at Socket.Readable.push (_stream_readable.js:135:10)
    at TCP.onread (net.js:542:20)
```

**Comment by javfindly** [comment]
same issue here, in my case happens when I run Jest:

node-debug node_modules/jest-cli/bin/jest.js -i

Node Inspector v0.12.8
Visit http://127.0.0.1:8080/?port=5858 to start debugging.
Debugging `node_modules/jest-cli/bin/jest.js`

Debugger listening on [::]:5858

/usr/local/lib/node_modules/node-inspector/lib/InjectorClient.js:112
      cb(error, NM[0].ref);
                     ^
TypeError: Cannot read property 'ref' of undefined
    at InjectorClient.<anonymous> (/usr/local/lib/node_modules/node-inspector/lib/InjectorClient.js:112:22)
    at /usr/local/lib/node_modules/node-inspector/lib/DebuggerClient.js:121:7

I've already checked, the NM object is an empty array

**Comment by maherma-adg** [comment]
I've the same issue, exception are thrown when connecting browser to :8080 port.

I think problem is related to node version, all works with node@6.3.1 and start failing when updated to node@6.4.0

All start working again after rolling back node to 6.3.1

**Comment by pjww** [comment]
I landed here after spending several hours trying to make things working.

Windows, Node 6.4.0 - x64 version.

When, after running node-inspector (or node-debug, or node --debug-brk, etc) I tried to go to the given address (with the GUI debugger with those devtools), the node proccess (and node-inspector proccess, if it was run separately) was crashing. No matter what I tried to do to solve it.

In the cmd, the stack trace wasn't really helpful (some exceptions from events.js and something about TCP, sorry, didn't save them).

From the browser, I could only learn that it had something to do with closed websocket and the inspector didn't worked at all (it was empty and console didn't work as well).

After using lynx1983's fix, I also noticed:

"No NativeModule in target scope."

After all, I downgraded Node to 6.3.1 and it's working.

**Comment by LeandroFavero** [comment]
Same problem

**Comment by brannon** [comment]
Not necessarily a good long term fix, but I worked around the problem with the following change to `node-inspector/lib/InjectorClient.js`:

```
108       if (!NM.length) {
109         error = new Error('No NativeModule in target scope');
110         return cb(error);
111       }
```

The debugger loads successfully and I am able to debug my code.

**Comment by polkovnikov-ph** [comment]
:+1: 

Downgrade to 6.3.1 helps. This is a regression,

**Comment by ding444** [comment]
For anyone using homebrew and trying to get back to node 6.3.1 after you've already done a cleanup you can re-install 6.3.1 with:

```sh
brew install https://github.com/Homebrew/homebrew-core/raw/cc180c23b583487101a1b6697da4126739948dc5/Formula/node.rb
```

and then switch with:

```sh
brew switch node 6.3.1
```

**Comment by bertolo1988** [comment]
> Node Inspector v0.12.8
> Visit http://127.0.0.1:8080/?port=5858 to start debugging.
> Debugging `node`
> 
> Debugger listening on [::]:5858
>  C:\Program Files\nodejs\node.exe:1
> (function (exports, require, module, __filename, __dirname) { MZ�
>                                                                 ^
> SyntaxError: Unexpected token ILLEGAL
>     at Object.exports.runInThisContext (vm.js:76:16)
>     at Module._compile (module.js:513:28)
>     at Object.Module._extensions..js (module.js:550:10)
>     at Module.load (module.js:458:32)
>     at tryModuleLoad (module.js:417:12)
>     at Function.Module._load (module.js:409:3)
>     at Timeout.Module.runMain [as _onTimeout] (module.js:575:10)
>     at tryOnTimeout (timers.js:228:11)
>     at Timer.listOnTimeout (timers.js:202:5)

This is also happening for node 6.3.1.

**Comment by thombaynes** [comment]
Small addition to the above code - I had to unlink node first, then I could install and switch to 6.3.1.
`brew unlink node`

**Comment by ELLIOTTCABLE** [comment]
Strangely, I see this with Node `v6.2.1`?

Meanwhile, has anybody tried @lynx1983's [patch](https://github.com/lynx1983/node-inspector/commit/3a26df74) referenced above?

**Comment by LeandroFavero** [comment]
Same problem in 6.5.0

**Comment by FJplant** [comment]
I succeeded with this ticket.

Simply, 
yum remove nodejs
yum install nodejs-6.3.1
worked for me. I use OEL 6.7 ( derivation of Redhat Enterprise Linux )

**Comment by aknuds1** [comment]
Also fails for me. Node 6.5.0 on OS X.

**Comment by aknuds1** [comment]
Maybe this issue should be split into two, as it addresses two separate bugs? One is the bad handling of the error case, the error being that NativeModule isn't found. PR #914 fixes the first one, so it would definitely help to have an issue dedicated to the second bug.

**Comment by kmkr** [comment]
Those running 6.5.0 can use native node for the same purposes. It's an experimental feature, but it works fine for me.

`node --inspect --debug-brk my-app.js`.

Read more [here](https://nodejs.org/api/debugger.html#debugger_v8_inspector_integration_for_node_js).

**Comment by gae123** [comment]
Thanks @kmkr but --inspect appears broken if you use the cluster API to fork workers, beware...

**Comment by wzup** [comment]
@kmkr

No they can't. `node --inspect` doesn't work on Windows 64 yet

**Comment by timothyerwin** [comment]
@wzup solution: stop using windows.

**Comment by odnarb** [comment]
Same issue.
OS: Ubuntu 15.10

```sh
brandon@ubuntu:~$ uname -a
Linux ubuntu 4.2.0-35-generic #40-Ubuntu SMP Tue Mar 15 22:15:45 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
brandon@ubuntu:~$ cat /etc/*-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=15.10
DISTRIB_CODENAME=wily
DISTRIB_DESCRIPTION="Ubuntu 15.10"
NAME="Ubuntu"
VERSION="15.10 (Wily Werewolf)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 15.10"
VERSION_ID="15.10"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
```
Also using node 6.5.0 & npm 3.10.3 with linux brew. My colleague is using Ubuntu as well but nvm with same node/npm releases as mine.  Same issue.

The hack @brannon suggested works for me for now.

**Comment by santanu-biswas** [comment]
Same problem with Node 6.5.0 on OSX El Capitan (Ver 10.11.6).
Downgrading to 6.3.1 did not solve the problem!!

```
Node Inspector v0.12.8
Visit http://127.0.0.1:8080/?port=5858 to start debugging.
/usr/local/lib/node_modules/node-inspector/lib/InjectorClient.js:111
      cb(error, NM[0].ref);
                     ^

TypeError: Cannot read property 'ref' of undefined
    at InjectorClient.<anonymous> (/usr/local/lib/node_modules/node-inspector/lib/InjectorClient.js:111:22)
    at /usr/local/lib/node_modules/node-inspector/lib/DebuggerClient.js:121:7
    at Object.create.processResponse.value (/usr/local/lib/node_modules/node-inspector/lib/callback.js:23:20)
    at Debugger._processResponse (/usr/local/lib/node_modules/node-inspector/lib/debugger.js:95:21)
    at Protocol.execute (_debugger.js:121:14)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:177:18)
    at Socket.Readable.push (_stream_readable.js:135:10)
    at TCP.onread (net.js:542:20)
```

**Comment by ellerbrock** [comment]
same problem here with:

```
Environment Information
========================
npm:   		 3.10.7
node:  		 v6.5.0
electron:      	 v1.3.5

MongoDB shell version: 3.2.8
Date: Mo 12 Sep 2016 03:45:28 ICT
```
i run node via nvm, with version 4x and 5x its working:

```
ok        Node v4.5.0
ok        Node v5.12.0
crash   Node v6.5.0
```

but crashes on current 6x with

```
▶ node-debug oop.js
Node Inspector v0.12.8
Visit http://127.0.0.1:8080/?port=5858 to start debugging.
Debugging `oop.js`

Error: listen EADDRINUSE :::5858
    at Object.exports._errnoException (util.js:1026:11)
    at exports._exceptionWithHostPort (util.js:1049:20)
    at Agent.Server._listen2 (net.js:1253:14)
    at listen (net.js:1289:10)
    at net.js:1399:9
    at _combinedTickCallback (internal/process/next_tick.js:77:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

**Comment by bhevesi** [comment]
node-inspector is the most popular debug tool for Node. Based on statistic, more than 10k downloads in the last day.
This open critical issue prevents using node v6.5.0 for developers.
I really appreciate Your debug framework and all effort by node-inspector developers but more than 29 days delay to provide solution is not fair. I'm sorry!

**Comment by CarlosLanderas** [comment]
Same issue here: 

Node version: v6.5.0
Npm version: 3.10.6
Node Inspector v0.12.8


 cb(error, NM[0].ref);
                     ^
TypeError: Cannot read property 'ref' of undefined

**Comment by e-cloud** [comment]
This is blocking issue!!! but seems like the team doesn't notice it. @3y3

**Comment by loic294** [comment]
Same here Windows x64
Node : 6.6.0

Downgrade is not a solution for me! Hope it's gone be fixed soon...

**Comment by tarol** [comment]
same as win10
node v6.5.0

**Comment by LeandroFavero** [comment]
The lastest release was in April 8. I don't know the future about node-inspector, but I believe that will die. Because NodeJS incorporated the inspector. Try this using 6.6.0:
node --inspect index.js

It's the same, but integrated and faster, really faster!!!

**Comment by polkovnikov-ph** [comment]
@LeandroFavero The problem is that hype on Node.js is over, and a lot of temporary solutions became permanent due to the lack of professional developers. Unfortunately `--inspect` is unstable and doesn't work for most of the interested users. I would expect `node-inspector` to disappear the same time as `node-gyp` would get Unicode support and drop Python from dependencies.

For the future commenters: consider using emoji reactions if you have the same setup as previous reporters. There're 2 messages a day with the same content, and it's annoying.

**Comment by timothyerwin** [comment]
@polkovnikov-ph why are you saying stuff like that? lack of professional developers? what is that based on? `--inspect` is new so they still need to patch it up. yes, it sucks that debugging isn't working out right now but they'll get it straightened out. it's unstable because the hype is real.

**Comment by ellerbrock** [comment]
the hype is real and [![Javascript](https://badges.frapsoft.com/javascript/code/javascript.svg?v=101)](https://github.com/ellerbrock/javascript-badges/) is here to stay.
this fact is based on true ♥

**Comment by fcvidoto** [comment]
im using windows 10 with node 6.6(before) ..

to solve this I uninstalled node and 
downloaded node 4.5 from the official nodejs site 

hope this helps someone

**Comment by bbones** [comment]
node 6.6.0
mocha 2.5.3
same error...

**Comment by adrian-moisa** [comment]
Same issue
win10
node v6.5.0

**Comment by fignuts** [comment]
I also have the same issue on win10
node v6.6.0

**Comment by polkovnikov-ph** [comment]
@fignuts but what version of node.js are you using?

**Comment by hlandao** [comment]
same issue here, running in OSX, node v6.5.0

**Comment by code-guru** [comment]
Same issue on Debain,
Node.js: v6.4.0
NPM: 3.10.3
Node Inspector v0.12.8

```
/../.nvm/versions/node/v6.6.0/lib/node_modules/node-inspector/lib/InjectorClient.js:111
      cb(error, NM[0].ref);
                     ^

TypeError: Cannot read property 'ref' of undefined
    at InjectorClient.<anonymous> (/.../.nvm/versions/node/v6.6.0/lib/node_modules/node-inspector/lib/InjectorClient.js:111:22)
```

It works on node v6.3.1

**Comment by hgwood** [comment]
Just in case someone misses it; #906 fixes this problem. You can patch your installed copy to get it working while waiting for the next release.

**Comment by e-cloud** [comment]
@hgwood, nope, it doesn't fix the real problem. It just fix the uncatch error to output it properly. If you apply the fix, node-inspector will throw another error(`No NativeModule in target scope`)

**Comment by hgwood** [comment]
Well yes, it does throw that error, but only once it has started, in its console. I can still step through my code, which I could not do before.

**Comment by e-cloud** [comment]
In my case, with that error no more debugging. I don't know what's the problem though :(.

**Comment by vird** [comment]
See hack/workaround solution https://github.com/node-inspector/node-inspector/issues/928#issuecomment-249343838

**Comment by yinhongyan** [comment]
Comment out line 111, it works!
![image](https://cloud.githubusercontent.com/assets/1445770/18808394/8686a2b6-8293-11e6-8250-7db3626d21f5.png)

**Comment by rolanday** [comment]
Same issue here -- node 6.7.0. Commenting out line 111 gets thinks working again as workaround. Thanks for suggesting!!

**Comment by benduran** [comment]
Tried upgrading to `6.7.0` in hopes this would be resolved. Still broken. `6.3.1` is the last working version for me on Win 10 Anniversary.

**Comment by WebReflection** [comment]
I don't understand why or how downgrading node version could solve the issue because [this code](https://github.com/node-inspector/node-inspector/blob/master/lib/InjectorClient.js#L108-L111) is clearly bugged.

Of course if `NM.length` is `0` any access to its indexes like `NM[0].ref` would fail.

Accordingly, before even thinking about a possible PR, what should actually happen when `NativeModule` is not found? A warning? Another attempt? Would passing the error and just `null` be appropriate?

**Comment by jeduden** [comment]
Instead of commenting out. Fixed the issue with using:
`cb(error, NM[0] && NM[0].ref);`

**Comment by WebReflection** [comment]
I've done the same but I don't know side effects, which is why I've asked
what should happen. Also a PR would be better

**Comment by 1j01** [comment]
Yeah, it should probably callback with the error (and `null`), that looks like the intention.

**Comment by raiteki** [comment]
Broken here too. Node 6.7.0 + MacOS Sierra. 

Debugger listening on [::]:5858
/usr/local/lib/node_modules/node-inspector/lib/InjectorClient.js:111
      cb(error, NM[0].ref);
                     ^

TypeError: Cannot read property 'ref' of undefined

**Comment by MasterJames** [comment]
Here is a Pull Request to patch this ASAP
https://github.com/node-inspector/node-inspector/pull/938

**Comment by WebReflection** [comment]
I've done a PR too but apparently nobody is maintaining this repo.

**Comment by hgwood** [comment]
Note that some (all?) node-inspector features are now built into Node: https://github.com/nodejs/node/pull/6792. See the `--inspect` command-line option.

**Comment by polkovnikov-ph** [comment]
@WebReflection It's a bit strange to agree with you, but you're totally correct. It was at least 2 months without a word from a maintainer. The last commit was in April. Looks like native Node's support is anticipated for a long time.

**Comment by joelkoz** [comment]
Also - add the "--debug-brk" option to "--inspect" and the debugger will pause execution until after you connect:

`node --inspect --debug-brk your.js`

You can then start execution from inside the debugger (after setting any necessary breakpoints).

**Comment by colthreepv** [comment]
Using `node --inspect` is not really a solution for everyone, for instance I am trying to debug an already alive process sending it a `USR1` signal to activate debug port

AFAIK this is not supported yet.

**Comment by miroslawmajka** [comment]
Beating the dead horse here but we had to go back to Node v.4.6.1 to get node-inspector working again. This is the setup:

Windows 10
Node v.4.6.1
Npm: 2.15.9
Node Inspector v.0.12.8

This setup works as of today and unless there is an explicit message in the Node or the Inspector changlog one day saying that they fixed this spcific issue we are sticking with this setup. Here's the download link for the old Node if anyone wants it: [v.4.6.1](https://nodejs.org/dist/v4.6.1/node-v4.6.1-x64.msi).

**Comment by WebReflection** [comment]
@miroslawmajka FYI `6.3.1` should work too. It's the latest version of node 6 that wasn't instantly crashing.

**Comment by leo-lai** [comment]
I also have the same issue on win7
node v6.7.0

**Comment by hilongjw** [comment]
I got seem issue with macOS and Nodejs 6.9.1
```
/usr/local/lib/node_modules/node-inspector/lib/InjectorClient.js:111
      cb(error, NM[0].ref);
                     ^

TypeError: Cannot read property 'ref' of undefined
    at InjectorClient.<anonymous> (/usr/local/lib/node_modules/node-inspector/lib/InjectorClient.js:111:22)
    at /usr/local/lib/node_modules/node-inspector/lib/DebuggerClient.js:121:7
    at Object.create.processResponse.value (/usr/local/lib/node_modules/node-inspector/lib/callback.js:23:20)
    at Debugger._processResponse (/usr/local/lib/node_modules/node-inspector/lib/debugger.js:95:21)
    at Protocol.execute (_debugger.js:121:14)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:172:18)
    at Socket.Readable.push (_stream_readable.js:130:10)
    at TCP.onread (net.js:542:20)
```

**Comment by simonbaker** [comment]
me too.
我安装了node 6.3.1版本就好了。。。

**Comment by Bargamut** [comment]
Hi!

It's my first steps in Node.js. =j
I've the same issue:

```
node --debug server.js
node-inspector
```
Node 6.9.2
and
Node 7.2.1

On the page:
> "Detached from the target by reason: websocket_closed"

In console:
```
/usr/local/lib/node_modules/node-inspector/lib/InjectorClient.js:111
      cb(error, NM[0].ref);
                     ^

TypeError: Cannot read property 'ref' of undefined
    at InjectorClient.<anonymous> (/usr/local/lib/node_modules/node-inspector/lib/InjectorClient.js:111:22)
    at /usr/local/lib/node_modules/node-inspector/lib/DebuggerClient.js:121:7
    at Object.create.processResponse.value (/usr/local/lib/node_modules/node-inspector/lib/callback.js:23:20)
    at Debugger._processResponse (/usr/local/lib/node_modules/node-inspector/lib/debugger.js:95:21)
    at Protocol.execute (_debugger.js:121:14)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:172:18)
    at Socket.Readable.push (_stream_readable.js:130:10)
    at TCP.onread (net.js:542:20)
```

**PS:**
**node -inspect server.js**  - success:

```
node --inspect echoserver.js 
Debugger listening on port 9229.
Warning: This is an experimental feature and could change at any time.
To start debugging, open the following URL in Chrome:
    chrome-devtools://devtools/bundled/inspector.html?experiments=true&v8only=true&ws=127.0.0.1:9229/10f1f810-c0fb-4ca4-aa31-77b221d326e8
```

**Comment by maithreya** [comment]
node: v6.9.1
osx: 10.12.2 sierra

![screen shot 2016-12-21 at 11 37 06 pm](https://cloud.githubusercontent.com/assets/3530852/21418537/c9fd87b6-c7d6-11e6-87d2-5131633a1276.png)
![screen shot 2016-12-21 at 11 37 25 pm](https://cloud.githubusercontent.com/assets/3530852/21418536/c9fb131e-c7d6-11e6-950e-deaf9591936e.png)


It works for me 💯
